js/liqui.js

```diff
@@ -508,35 +508,31 @@ module.exports = class liqui extends Exchange {
         return this.orders[id];
     }
 
-    async fetchOrders (symbol = undefined, since = undefined, limit = undefined, params = {}) {
-        // if (!symbol)
-        //     throw new ExchangeError (this.id + ' fetchOrders requires a symbol');
-        await this.loadMarkets ();
-        let request = {};
-        let market = undefined;
-        if (symbol) {
-            let market = this.market (symbol);
-            request['pair'] = market['id'];
-        }
-        let response = await this.privatePostActiveOrders (this.extend (request, params));
-        let openOrders = [];
-        if ('return' in response)
-            openOrders = this.parseOrders (response['return'], market);
+    updateOrdersCache (openOrders, symbol) {
         for (let j = 0; j < openOrders.length; j++) {
-            this.orders[openOrders[j]['id']] = openOrders[j];
+            // update local cache with opened orders
+            const id = openOrders[j]['id'];
+            this.orders[id] = openOrders[j];
         }
         let openOrdersIndexedById = this.indexBy (openOrders, 'id');
         let cachedOrderIds = Object.keys (this.orders);
-        let result = [];
         for (let k = 0; k < cachedOrderIds.length; k++) {
+            // try to find every cached order in open orders array
+            // possible reasons why cached order is not there:
+            // - order was closed or cancelled -> update cache
+            // - symbol mismatch (e.g. cached BTC/USDT, fetched ETH/USDT) -> skip
             let id = cachedOrderIds[k];
-            if (id in openOrdersIndexedById) {
-                this.orders[id] = this.extend (this.orders[id], openOrdersIndexedById[id]);
-            } else {
+            if (!(id in openOrdersIndexedById)) {
+                // cached order is not in open orders array
                 let order = this.orders[id];
+                if (typeof symbol !== 'undefined' && symbol !== order['symbol']) {
```

Comment:
I'm totally ok with merging this pull request, and wanted to thank you again for your participation in this project.

This is not an issue nor a problem at all, just wanted to share a couple of thoughts. When I'm looking at this line, here's what I'm thinking of: we are lucky that it does the job for Liqui, as their order ids (id-spaces) rarely intersect. However, I can foresee problems with this type of design if we run into an exchange that has symbol-specific ids, like you said earlier before... And from here, to make this method really generic we might need two types of cache:
- an *orders indexed by id* cache (current type)
- an *orders grouped by symbol indexed by id* cache (missing now)

Again these thought are not related to the pull request, I'm merging it very soon. But apparently this method is going to migrate to the base class someday, so, we might take a few considerations along the way )

Reply:
Yeah, it's a good catch, as you would say )

By their nature orders belong to markets even if they share exchange-wide id space. And if we look at it from this perspective I think we won't need 'orders indexed by id' at all, 'orders grouped by symbol indexed by id' are generic enough to serve all cases, right?